design/ErrorHandling.md

```diff
@@ -80,90 +79,244 @@ device.addEventListener('validationerror', (event) => {
 
-### Error propagation of invalid objects
+### Categories of WebGPU Calls
 
-Using any invalid object in a WebGPU operation produces a validation error.
-The effect of an error depends on the type of a call:
+Each category of WebGPU call has its errors handled in a different way.
 
- - For object creation, the call produces a new invalid object.
- - For `GPUCommandEncoder` encoding methods, the `GPUCommandEncoder.finishEncoding` method will return an invalid object.
- - For other WebGPU calls, the call becomes a no-op.
+#### Initialization
 
-In each case, an error is logged to the error log.
+Creation of the adapter and device.
 
-## *Recovery*: Recoverable Errors
+  - `gpu.requestAdapter`
+  - `GPUAdapter.requestDevice`
 
-Recoverable errors are produced only by object creation.
-The status of an object can be retrieved asynchronously (see next section).
+Handled by "Fatal Errors" above.
 
-```
-enum GPUObjectStatus {
-    // The object is valid.
-    "valid",
-    // The object is invalid due to a non-fatal allocation failure.
-    // The application can use this as a signal to scale back resource usage, if possible.
-    "out-of-memory",
-    // The object is invalid for another, unrecoverable reason.
-    "invalid",
+#### Object-Returning
+
+WebGPU Object creation and getters.
+
+  - `GPUDevice.create*`
+  - `GPUTexture.create*View`
+  - `GPUCommandEncoder.finish`
+  - `GPUDevice.getQueue`
+  - `GPUSwapChain.getCurrentTexture`
+
+If there is an error, the returned object is invalid.
+
+See "Object Status" below.
+
+#### Object-Promise-Returning*
+
+Async WebGPU Object creation.
+
+  - `GPUDevice.createReady*`
+  - `GPUDevice.createBufferMappedAsync`
+
+If there is an error, either the returned object is invalid, or the Promise rejects.
+
+See "Object Status" below.
+
+#### Encoding
+
+Recording of GPU commands in `GPUCommandEncoder`.
+
+  - `GPU*Encoder.*`
+
+These commands do not report errors.
+Instead, any error is reported by `GPUCommandEncoder.finish`.
+
+#### Other-Promise-Returning
+
+  - `GPUDevice.getSwapChainPreferredFormat`
+  - `GPUFence.onCompletion`
+  - `GPUBuffer.mapReadAsync`
+  - `GPUBuffer.mapWriteAsync`
+
+If there is an error, the returned Promise rejects.
+
+#### Void-Returning
+
+  - `GPUQueue.submit`
+  - `GPUQueue.signal`
+  - `GPUBuffer.setSubData`
+  - `GPUBuffer.unmap`
+  - `GPUBuffer.destroy`
+  - `GPUTexture.destroy`
+  - `GPUFence.getCompletedValue`
+
+See "Operation Status" below.
+
+## Object Status
+
+As described above, objects may be internally "valid" or "invalid".
+For testing, it is necessary to be able to inspect this status.
+Through this mechanism, allocation failure ("out-of-memory") can also be expressed.
+
+Object status can be queried with `getStatus`.
+`getStatus` returns a Promise which resolves to a `GPUObjectStatus` object that reports the status of the object.
```

Comment:
Like with the rest of the API, I'm trying to avoid adding any synchronous stuff. What use case do you have in mind that would make it necessary to getObjectStatus synchronously?

E.g. for pose data, the app can use (1) an existing buffer with setSubData, (2) an existing, already mapped buffer, or (3) createBufferMapped

Reply:
I think we'll probably have a Dev Tools option to (synchronously) "break on WebGPU error" as the dev mode.

Reply:
Re: Use case:
The purpose of the PR is to add mechanisms whereby developers can easily debug their code when things go wrong. To that end, using the testing APIs should not change the behavior of the application with regard to the sequencing of RAFs in relation to one another or other events/callbacks running in the application.

Reply:
I am actually not trying to design these mechanisms for debugging, since Dev Tools would hopefully provide good tools.

But it's a good point that sometimes you do want more programmatic debugging tools (printf style). I'll keep it in mind.